# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/scene_building.rs
# Commit: fc2b3d6448bc
# Full Hash: fc2b3d6448bcc3575fffb1abc923e7e703aeef54
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2022-02-24 03:51:41
# Regressor Bug: 1749380
# File Overlap Count: 1
# Description:
#   Bug 1749380 - Part 2 - Performance and quality fixes for part 1. r=gfx-reviewers,nical
#   
#   * Add support for local scale factors to a surface, allowing it to
#     be rasterized in root coordinate space. This allows snapping to
#     work across surfaces where the surface transform is a fractional
# ==============================================================================

diff -r 8c70bc92ad36 -r fc2b3d6448bc gfx/wr/webrender/src/scene_building.rs
--- a/gfx/wr/webrender/src/scene_building.rs	Wed Feb 23 20:49:26 2022 +0000
+++ b/gfx/wr/webrender/src/scene_building.rs	Wed Feb 23 20:49:27 2022 +0000
@@ -229,6 +229,8 @@
     spatial_node_index: SpatialNodeIndex,
     /// Prim flags for any pictures in this chain
     flags: PrimitiveFlags,
+    /// Requested raster space for enclosing stacking context
+    raster_space: RasterSpace,
 }
 
 impl PictureChainBuilder {
@@ -237,6 +239,7 @@
         prim_list: PrimitiveList,
         flags: PrimitiveFlags,
         spatial_node_index: SpatialNodeIndex,
+        raster_space: RasterSpace,
     ) -> Self {
         PictureChainBuilder {
             current: PictureSource::PrimitiveList {
@@ -244,6 +247,7 @@
             },
             spatial_node_index,
             flags,
+            raster_space,
         }
     }
 
@@ -252,6 +256,7 @@
         instance: PrimitiveInstance,
         flags: PrimitiveFlags,
         spatial_node_index: SpatialNodeIndex,
+        raster_space: RasterSpace,
     ) -> Self {
         PictureChainBuilder {
             current: PictureSource::WrappedPicture {
@@ -259,6 +264,7 @@
             },
             flags,
             spatial_node_index,
+            raster_space,
         }
     }
 
@@ -300,6 +306,7 @@
                 self.flags,
                 prim_list,
                 self.spatial_node_index,
+                self.raster_space,
             ))
         );
 
@@ -316,6 +323,7 @@
             },
             spatial_node_index: self.spatial_node_index,
             flags: self.flags,
+            raster_space: self.raster_space,
         }
     }
 
@@ -344,6 +352,7 @@
                         self.flags,
                         prim_list,
                         self.spatial_node_index,
+                        self.raster_space,
                     ))
                 );
 
@@ -2073,6 +2082,7 @@
                 is_redundant,
                 is_backdrop_root: flags.contains(StackingContextFlags::IS_BACKDROP_ROOT),
                 flags,
+                raster_space: new_space,
             });
         }
 
@@ -2168,6 +2178,7 @@
                         stacking_context.prim_flags,
                         stacking_context.prim_list,
                         stacking_context.spatial_node_index,
+                        stacking_context.raster_space,
                     ))
                 );
 
@@ -2182,6 +2193,7 @@
                     instance,
                     stacking_context.prim_flags,
                     stacking_context.spatial_node_index,
+                    stacking_context.raster_space,
                 )
             }
             Picture3DContext::Out => {
@@ -2190,6 +2202,7 @@
                         stacking_context.prim_list,
                         stacking_context.prim_flags,
                         stacking_context.spatial_node_index,
+                        stacking_context.raster_space,
                     )
                 } else {
                     let composite_mode = Some(
@@ -2206,6 +2219,7 @@
                             stacking_context.prim_flags,
                             stacking_context.prim_list,
                             stacking_context.spatial_node_index,
+                            stacking_context.raster_space,
                         ))
                     );
 
@@ -2220,6 +2234,7 @@
                         instance,
                         stacking_context.prim_flags,
                         stacking_context.spatial_node_index,
+                        stacking_context.raster_space,
                     )
                 }
             }
@@ -2304,6 +2319,7 @@
                     stacking_context.prim_flags,
                     prim_list,
                     stacking_context.spatial_node_index,
+                    stacking_context.raster_space,
                 ))
             );
 
@@ -2318,6 +2334,7 @@
                 instance,
                 stacking_context.prim_flags,
                 stacking_context.spatial_node_index,
+                stacking_context.raster_space,
             );
         }
 
@@ -2797,6 +2814,7 @@
                                 PrimitiveFlags::IS_BACKFACE_VISIBLE,
                                 prim_list,
                                 pending_shadow.spatial_node_index,
+                                RasterSpace::Screen,
                             ))
                         );
 
@@ -3492,6 +3510,7 @@
                     prim_flags,
                     prim_list,
                     backdrop_spatial_node_index,
+                    stacking_context.raster_space,
                 ))
             );
 
@@ -3507,6 +3526,7 @@
             instance,
             info.flags,
             backdrop_spatial_node_index,
+            RasterSpace::Screen,
         );
 
         source = self.wrap_prim_with_filters(
@@ -3784,6 +3804,9 @@
 
     /// Flags identifying the type of container (among other things) this stacking context is
     flags: StackingContextFlags,
+
+    /// Requested raster space for this stacking context
+    raster_space: RasterSpace,
 }
 
 impl FlattenedStackingContext {
@@ -3861,6 +3884,7 @@
                 self.prim_flags,
                 mem::replace(&mut self.prim_list, PrimitiveList::empty()),
                 self.spatial_node_index,
+                self.raster_space,
             ))
         );
 